File: prompt_guard/detector.py
"""Utilities for identifying and masking sensitive information in prompts."""
from __future__ import annotations

from dataclasses import dataclass
import re
from typing import Iterable, List, Pattern, Sequence, Tuple


@dataclass(frozen=True)
class SensitivePattern:
    """Configuration for a sensitive information pattern."""

    label: str
    pattern: Pattern[str]


@dataclass(frozen=True)
class SensitiveMatch:
    """Represents a single sensitive segment detected in a prompt."""

    label: str
    value: str
    placeholder: str
    start: int
    end: int


class SensitiveInfoDetector:
    """Detects and masks sensitive entities inside prompts."""

    def __init__(self, patterns: Iterable[SensitivePattern]):
        patterns = tuple(patterns)
        if not patterns:
            raise ValueError("At least one SensitivePattern is required")
        self._patterns: Tuple[SensitivePattern, ...] = patterns

    @staticmethod
    def default_patterns() -> Tuple[SensitivePattern, ...]:
        """Return a curated set of default detection rules."""

        client_names = r"\b(?:Acme Corp|Globex|Initech|Umbrella|Hooli|Vehement Capital Partners)\b"
        kpi_values = (
            r"\b(?:revenue|ebitda|profit|conversion rate|churn rate|"
            r"net promoter score|ARR|customer acquisition cost)\b[^\n]*?\d[\w$%,.-]*"
        )
        card_like = r"\b(?:\d[ -]?){13,16}\b"
        email_like = r"[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}"
        return (
            SensitivePattern("client_name", re.compile(client_names, re.IGNORECASE)),
            SensitivePattern("kpi", re.compile(kpi_values, re.IGNORECASE)),
            SensitivePattern("possible_account_number", re.compile(card_like)),
            SensitivePattern("email", re.compile(email_like)),
        )

    def scan(self, text: str) -> List[SensitiveMatch]:
        """Scan ``text`` for sensitive information and return the matches."""

        spans: List[Tuple[int, int, SensitivePattern, str]] = []
        for pattern in self._patterns:
            for match in pattern.pattern.finditer(text):
                spans.append((match.start(), match.end(), pattern, match.group(0)))

        spans.sort(key=lambda item: (item[0], -(item[1] - item[0])))

        matches: List[SensitiveMatch] = []
        last_end = -1
        counter = 1
        for start, end, pattern, value in spans:
            if start < last_end:
                # Skip overlapping matches, keep the earliest occurrence.
                continue
            placeholder = f"{{{{SENSITIVE_{counter}}}}}"
            matches.append(
                SensitiveMatch(
                    label=pattern.label,
                    value=value,
                    placeholder=placeholder,
                    start=start,
                    end=end,
                )
            )
            counter += 1
            last_end = end
        return matches

    def mask(self, text: str) -> Tuple[str, List[SensitiveMatch]]:
        """Mask sensitive segments in ``text`` and return the sanitized prompt."""

        matches = self.scan(text)
        if not matches:
            return text, []

        result_parts = []
        last_idx = 0
        for match in matches:
            result_parts.append(text[last_idx:match.start])
            result_parts.append(match.placeholder)
            last_idx = match.end
        result_parts.append(text[last_idx:])

        sanitized = "".join(result_parts)
        return sanitized, matches

    def restore(self, text: str, matches: Sequence[SensitiveMatch]) -> str:
        """Restore previously masked placeholders with their original values."""

        restored = text
        for match in matches:
            restored = restored.replace(match.placeholder, match.value)
        return restored

File: vscode-extension/src/detector.ts
import { SensitiveMatch, SensitivePatternConfig } from "./types";

interface InternalPattern {
  label: string;
  regex: RegExp;
}

const BUILTIN_PATTERNS: InternalPattern[] = [
  {
    label: "client_name",
    regex: /\b(?:Acme Corp|Globex|Initech|Umbrella|Hooli|Vehement Capital Partners)\b/i,
  },
  {
    label: "kpi",
    regex:
      /\b(?:revenue|ebitda|profit|conversion rate|churn rate|net promoter score|ARR|customer acquisition cost)\b[^\n]*?\d[\w$%,.-]*/i,
  },
  {
    label: "possible_account_number",
    regex: /\b(?:\d[ -]?){13,16}\b/,
  },
  {
    label: "email",
    regex: /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/,
  },
  {
    label: "private_key",
    regex: /-----BEGIN (?:RSA |DSA |EC |OPENSSH |PGP )?PRIVATE KEY-----/,
  },
  {
    label: "cursor_api_key",
    regex: /(?:^|[^A-Za-z0-9_-])crsr_[A-Za-z0-9_-]{20,}(?:[^A-Za-z0-9_-]|$)/,
  },
  {
    label: "github_token",
    regex: /(?:^|[^A-Za-z0-9_])(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9_]{30,}(?:[^A-Za-z0-9_]|$)/,
  },
  {
    label: "slack_token",
    regex: /(?:^|[^A-Za-z0-9_-])xox[baprs]-[A-Za-z0-9-]{20,}(?:[^A-Za-z0-9_-]|$)/,
  },
  {
    label: "openai_api_key",
    regex: /(?:^|[^A-Za-z0-9_-])sk-[A-Za-z0-9_-]{32,}(?:[^A-Za-z0-9_-]|$)/,
  },
  {
    label: "aws_access_key",
    regex: /(?:^|[^A-Za-z0-9_])(?:AKIA|ASIA)[A-Z0-9]{16}(?:[^A-Za-z0-9_]|$)/,
  },
  {
    label: "jwt_token",
    regex:
      /(?:^|[^A-Za-z0-9_-])eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}(?:[^A-Za-z0-9_-]|$)/,
  },
  {
    label: "ssn",
    regex: /\b\d{3}-\d{2}-\d{4}\b/,
  },
  {
    label: "phone_number",
    regex: /\b(?:\+?1[-.\s]?)?(?:\(\d{3}\)|\d{3})[-.\s]?\d{3}[-.\s]?\d{4}\b/,
  },
];

function buildPatterns(customPatterns: SensitivePatternConfig[]): InternalPattern[] {
  const patterns = [...BUILTIN_PATTERNS];
  for (const custom of customPatterns) {
    try {
      patterns.push({
        label: custom.label,
        regex: new RegExp(custom.pattern, custom.ignoreCase ? "i" : undefined),
      });
    } catch {
      // Invalid regex in settings is ignored to keep the extension running.
    }
  }
  return patterns;
}

export function scanText(text: string, customPatterns: SensitivePatternConfig[] = []): SensitiveMatch[] {
  const patterns = buildPatterns(customPatterns);
  const spans: Array<{ start: number; end: number; label: string; value: string }> = [];

  for (const pattern of patterns) {
    const regex = new RegExp(pattern.regex.source, pattern.regex.flags + (pattern.regex.global ? "" : "g"));
    for (const match of text.matchAll(regex)) {
      if (match.index === undefined) {
        continue;
      }
      spans.push({
        start: match.index,
        end: match.index + match[0].length,
        label: pattern.label,
        value: match[0],
      });
    }
  }

  spans.sort((left, right) => left.start - right.start || right.end - right.start - (left.end - left.start));

  const matches: SensitiveMatch[] = [];
  let lastEnd = -1;
  let counter = 1;

  for (const span of spans) {
    if (span.start < lastEnd) {
      continue;
    }
    matches.push({
      label: span.label,
      value: span.value,
      placeholder: `{{SENSITIVE_${counter}}}`,
      start: span.start,
      end: span.end,
    });
    counter += 1;
    lastEnd = span.end;
  }

  return matches;
}

export function maskText(text: string, matches: SensitiveMatch[]): string {
  if (matches.length === 0) {
    return text;
  }

  const parts: string[] = [];
  let lastIndex = 0;
  for (const match of matches) {
    parts.push(text.slice(lastIndex, match.start));
    parts.push(match.placeholder);
    lastIndex = match.end;
  }
  parts.push(text.slice(lastIndex));
  return parts.join("");
}

export function formatMatchSummary(matches: SensitiveMatch[]): string {
  if (matches.length === 0) {
    return "Prompt Guard: no sensitive data detected";
  }
  const labels = [...new Set(matches.map((match) => match.label))];
  return `Prompt Guard: ${matches.length} sensitive segment(s) (${labels.join(", ")})`;
}
